diff/parse.go

```diff
@@ -372,6 +372,104 @@ func (r *FileDiffReader) ReadExtendedHeaders() ([]string, error) {
 	}
 }
 
+// readQuotedFilename extracts a quoted filename from the beginning of a string,
+// returning the unquoted filename and any remaining text after the filename.
+func readQuotedFilename(text string) (value string, remainder string, err error) {
+	if text[0] != '"' {
+		panic("caller must ensure filename is quoted! " + text)
+	}
+
+	// The end quote is the first quote NOT preceeded by an uneven number of backslashes.
+	var i, j int
+	for i = 1; i < len(text); i++ {
```

Comment:
Since the code deals strictly with ASCII characters, it can safely ignore UTF-8 sequences. Given the amount of string indices in play, I wanted to keep things simple.

However, having now rewritten the code to not walk backwards, the result is simple enough with a for range.